micro-rdk/src/common/ota.rs

```diff
@@ -352,53 +363,97 @@ impl<S: OtaMetadataStorage> OtaService<S> {
                                 .handshake(io)
                                 .await
                             {
-                                Ok(pair) => break pair,
+                                Ok(pair) => {
+                                    sender = Some(pair.0);
+                                    inner_conn = Some(pair.1);
+                                    retry = false;
+                                }
                                 Err(e) => {
                                     log::error!("failed to build http request: {}", e);
+                                    retry = true;
                                 }
                             }
                         }
                         Err(e) => {
                             log::error!("failed to create tcp stream: {}", e);
+                            retry = true;
                         }
                     }
                 }
                 Err(e) => {
                     log::error!("failed to create http connection: {}", e);
+                    retry = true;
                 }
             };
 
-            log::debug!(
-                "retry establishing http connection to {} in {} seconds",
-                &uri,
-                CONN_RETRY_SECS
-            );
-            Timer::after(Duration::from_secs(CONN_RETRY_SECS)).await;
-        };
-
-        let conn = Box::new(self.exec.spawn(async move {
-            if let Err(err) = conn.await {
-                log::error!("connection failed: {:?}", err);
+            if sender.is_none() || inner_conn.is_none() {
+                return Err(OtaError::Other(
+                    "invalid state, error path not handled".to_string(),
+                ));
+            }
+            let mut sender = sender.unwrap();
+            let inner_conn = inner_conn.unwrap();
+
+            if retry {
+                log::debug!(
```

Comment:
[optional] - I'd consider having this at `info` level, if not `warn`, with a somewhat more polished message.